RFC for automatically generated typescript API documentation for every plugins public services, types, and functionality (#86704)
elastic/kibana: rfcs/text/0014_api_documentation.md

single json file easier.

```ts
export interface ScopeApi {
  setup?: ApiDeclaration;
  start?: ApiDeclaration;
  classes: ApiDeclaration[];
  functions: ApiDeclaration[];
  interfaces: ApiDeclaration[];
  objects: ApiDeclaration[];
  enums: ApiDeclaration[];
  misc: ApiDeclaration[];
  // We may add more here as we sit fit to pull out of `misc`.
}
```

With this structure, the mdx files end up looking like:

```
### Start
<DocDefinitionList data={[actionsJson.server.start]}/>
### Functions
<DocDefinitionList data={actionsJson.server.functions}/>
### Interfaces
<DocDefinitionList data={actionsJson.server.interfaces}/>
```

### PluginApi

A plugins API is the component that is serialized into the json file. It is broken into public, server and common components.  `serviceFolders` is a way for the system to
write separate mdx files depending on where each declaration is defined. This is because certain plugins (and core)
are huge, and can't be rendered in a single page.

```ts
export interface PluginApi {
  id: string;
  serviceFolders?: readonly string[];
  client: ScopeApi;
  server: ScopeApi;
  common: ScopeApi;
}
```

## Manifest file changes

**serviceFolders?: string[]**

> Used by the system to group services into sub-pages. Some plugins, like data and core, have such huge APIs they are very slow to contain in a single page, and they are less consummable by solution developers. The addition of an optional list of services folders will cause the system to automatically create a separate page with every API that is defined within that folder. The caveat is that core will need to define a manifest file in order to define its service folders...

Author: I could look into making this work. Do we use namespaces anywhere? I thought we discouraged using them for a bit.  The data team is using an Object to simulate a namespace, which isn't great because I can't distinguish between an object that is used only to namespace and an object that is meant to be exported as an actual object.

> Used by the system to group services into sub-pages. Some plugins, like data and core, have such huge APIs they are very slow to contain in a single page, and they are less consummable by solution developers. The addition of an optional list of services folders will cause the system to automatically create a separate page with every API that is defined within that folder. The caveat is that core will need to define a manifest file in order to define its service folders...

Author: I should mention as well though, this sounds like a lot of work that would hold up a V1 of docs, so I would opt to merge using the folder approach and then improve it. The docs are unusable with core and data plugins as big as they are, so I can't merge as a single page. Even just hard coding some plugin splits, in order to get a v1 docs API merged, then iterated. 

I thought the folder approach would also be nice to mimic our "organize by domain" folder structure principal. Having the API reflect the folder structure could be helpful for discoverability.

> Used by the system to group services into sub-pages. Some plugins, like data and core, have such huge APIs they are very slow to contain in a single page, and they are less consummable by solution developers. The addition of an optional list of services folders will cause the system to automatically create a separate page with every API that is defined within that folder. The caveat is that core will need to define a manifest file in order to define its service folders...

Reviewer: yeah introducing namespaces will require refactoring a lot of downstream types so I agree, going with `serviceFolders` is a better first step.

**Using a kibana.json file for core**

For the purpose of API infrastructure, core is treated like any other plugin. This means it has to specify serviceFolders section inside a manifest file to be split into sub folders. There are other ways to tackle this - like a hard coded array just for the core folder, but I kept the logic as similar to the other plugins as possible.

**teamOwner: string**

Team owner can be determined via github CODEOWNERS file, but we want to encourage single team ownership per plugin. Requiring a team owner string in the manifest file will help with this and will allow the API doc system to manually add a section to every page that has a link to the team owner. Additional ideas are teamSlackChannel or teamEmail for further contact. 

**summary: string**

A brief description of the plugin can then be displayed in the automatically generated API documentation.

## Technology: ts-morph vs api-extractor

> [Api-extractor](https://api-extractor.com/) is a utility built from microsoft that parses typescript code into json files that can then be used in a custom [api-documenter](https://api-extractor.com/pages/setup/generating_docs/) in order to build documentation. This is what we [have now](https://github.com/elastic/kibana/tree/master/docs/development), except we use the default api-documenter. Unfortunately, because our plugins aren’t separate packages, and the way api-extractor works, we can’t build cross plugin links this way.

Reviewer: Since api-extractor-model supports loading several api.json files the API theoretically supports resolving cross-plugin links. We currently use kibana's top-level package.json file when running api-extractor so all api.json files use the "kibana" package name. If instead of supplying a `packageJsonFullPath` we supplied a generated `packageJson` with the name of the plugin we would get the correct package names in the generated api.json files https://github.com/elastic/kibana/blob/master/src/dev/run_check_published_api_changes.ts#L101

Author: I'll add that point to this section. I think I tried to explore this approach but wasn't able to get to it working. It's fully possible I overlooked something, or overestimated the effort/underestimated the feasibility. 

I still think it's worth trying out the current approach with ts-morph, and re-evaluating as time goes on.

It is unknown whether the work with bazel will allow us to use this technology. I suspect not, because even though there will be separate `.d.ts` files, the plugins will still be importing types via relative links.

[ts-morph](https://github.com/dsherret/ts-morph) is a utility built and maintained by a single person, which sits a layer above the raw typescript compiler. It affords greater flexibility, thus supports cross plugin links (among other things like links to source files). The downsides of using this library are:

 - Risks of relying on a package maintained by a single developer

>  - Less re-usability across repositories. What if EUI wanted to use the same system?  

Author: I'll note your concern in the RFC, I agree it's an important one.

The author does seem to regularly keep up with typescript upgrades, but there is still risk that he will slow down, or certain upgrades will require a large effort on our part.

> In my opinion more expensive than not having docs, but this is hard to quantify

We always have the option of making this call in the future. Each time we prepare for a typescript upgrade, we'll need to decide between:

1. Support the new version in our ts-morph implementation 
2. Switch to an api-extractor implementation
3. Remove our API docs

If it's easy to upgrade ts-morph, we make the changes and move on. If it's not, we assess how useful we think the docs are compared to the effort it would take to keep them. If the effort outweighs their usefulness, we can remove them, perhaps temporarily, until there is more bandwidth to fix the issues and bring back support.

If you're worried about devs starting to rely on docs that we might need to stop supporting in order to move faster with typescript, I think that would indicate a high level of importance, and mean it's worth investing the effort into keeping them around.

> I recommend we move ahead with ts-morph, acknowleding the possibility of migrating to api-extractor in the future. If so, the effort shouldn’t be a large one, either way we will need logic to transform data into ApiDeclarations, it will either be ts-morph nodes, or api-extractor json output.

Author: That is _very_ fair. When I was re-reading this today, I felt the same way. If it wasn't a lot of effort, I would have done it already.  I've reworded this section.

parameters, not as much when thinking about an exported variable that might be undefined.

- Signature. This is only relevant for some types: functions, objects, type, arrays and compound. Classes and interfaces would be too large.
For primitives, this is equivalent to "type".

- Children. Only relevant for some types, this would include parameters for functions, class members and functions for classes, properties for
interfaces and objects. This makes the structure recursive. Each child is a nested API component.

- Return comment. Only relevant for function types.

![image](../images/api_info.png)

### ApiDeclaration type

```ts
interface ApiDeclaration {
  label: string;
  type: TypeKind; // string, number, boolean, class, interface, function, type, etc.
  description: TextWithLinks;
  signature: TextWithLinks;
  tags: string[];  // Declarations may be tagged as beta, or deprecated.
  children: ApiDeclaration[]; // Recursive - this could be function parameters, class members, or interface/object properties.
  returnComment?: TextWithLinks;
  lifecycle?: Lifecycle.START | Lifecycle.SETUP;
}

```

# Architecture design

## Location

> The generated docs will reside inside the kibana repo, inside a top level `api_docs` folder. In the long term, we could investigate having the docs system run a script to generated the mdx files, so we don’t need to store them inside the repo. Every ci run should destroy and re-create this folder so removed plugins don't have lingering documentation files.

Author: I don't think so - I always found it really weird that we have `docs/developer` and `docs/development` and one is for autotgenerated md files and the other is hand written asciidoc. 

If we could split `docs` into two sub folders  (one hand written, the other auto generated) I think that would be okay, but that would be a lot work. Note we also now have `dev_docs` which I'm using for high level, cross plugin documentation (hand written).

> The generated docs will reside inside the kibana repo, inside a top level `api_docs` folder. In the long term, we could investigate having the docs system run a script to generated the mdx files, so we don’t need to store them inside the repo. Every ci run should destroy and re-create this folder so removed plugins don't have lingering documentation files.

Reviewer 2: > I don't think so - I always found it really weird that we have docs/developer and docs/development and one is for autotgenerated md files and the other is hand written asciidoc.

Agreed that the current folder-structure is super awkward.

> If we could split docs into two sub folders (one hand written, the other auto generated) I think that would be okay, but that would be a lot work. Note we also now have dev_docs which I'm using for high level, cross plugin documentation (hand written).

Maybe we leave `docs` as the hand-written docs, and add a `generated_docs` folder with `plugin_api` as the child? I'm primarily trying to prevent us from continuing to add `${something}_docs` folders because this will soon become unwieldy.

> The generated docs will reside inside the kibana repo, inside a top level `api_docs` folder. In the long term, we could investigate having the docs system run a script to generated the mdx files, so we don’t need to store them inside the repo. Every ci run should destroy and re-create this folder so removed plugins don't have lingering documentation files.

Author: Resolved offline. I'll continue with the current approach, but will restructure if any more top level doc folders show up, or once the current `docs` folder no longer contains asciidoc.

    children: getParameterList(node.getParameters(), getParamTagComments(node)),
  }
} if (...) 
....
```

The handling of each specific type is what encompasses the vast majority of the logic in the PR.

The public and server scope have 0-2 special interfaces indicated by "lifecycle". This is determined by using ts-morph to extract the first two generic types 
passed to `... extends Plugin<start, setup>` in the class defined inside the plugin's `plugin.ts` file.

A [PluginApi](#pluginapi) is generated for each plugin, which is used to generate the json and mdx files. One or more json/mdx file pair
 per plugin may be created, depending on the value of `serviceFolders` inside the plugin's manifest files. This is because some plugins have such huge APIs that
 it is too large to render in a single page.

![image](../images/api_doc_tech.png)

## Types

### TypeKind

TypeKind is an enum that will identify what "category" or "group" name we can call this particular export. Is it a function, an interface, a class a variable, etc.
This list is likely incomplete, and we'll expand as needed.

```ts
export enum TypeKind {

>   ClassKind = 'Class',

Author: Nope, I can make them all post fixed by `Kind`.

  FunctionKind = 'Function',
  ObjectKind = 'Object',
  InterfaceKind = 'Interface',
  TypeKind = 'Type', // For things like `export type Foo = ...`

>   Unknown = 'Unknown', // There are a lot of ts-morph types, if I encounter something not handled, I dump it in here.

Author: Oh good thought.  yes, `any` falls under here.  

<img width="594" alt="Screen Shot 2021-02-01 at 5 22 05 PM" src="https://user-images.githubusercontent.com/16563603/106532965-3a06fe00-64bf-11eb-9c9b-209f98053148.png">

Hmmm... I could add `any` and `unknown` as TypeKinds, and come up with a new name for "don't know what this is". "Uncategorized"?

>   Unknown = 'Unknown', // There are a lot of ts-morph types, if I encounter something not handled, I dump it in here.

Reviewer 2: > Hmmm... I could add any and unknown as TypeKinds, and come up with a new name for "don't know what this is". "Uncategorized"?

"Uncategorized" seems reasonable. We could also make `TypeKind` nullable/undefined. I think either is fine.

  String = 'string',
  Number = 'number',
  Boolean = 'boolean',
  Array = 'Array',
  CompoundType = 'CompoundType', // Unions & intersections, to handle things like `string | number`. 
}
```

### Text with reference links

Signatures, descriptions and returns comments may all contain links to other API declarations. This information needs to be serializable into json. This serializable type encompasses the information needed to build the DocLink components within these fields. The logic of building

> the DocLink components currently resides inside the elasticdocs system. It's unclear if this will change.

Reviewer 2: super-nit
```suggestion
the DocLink components currently resides inside the elastic-docs system. It's unclear if this will change.
```

 *   ') => void'
 * ]
 * ```
 *
 * This is then used to render text with nested DocLinks so it looks like this:
 *
 * `(a: => <DocLink docId="pluginB" section="Bar" text="Bar"/>) => void`
 */
export type TextWithLinks = Array<string | Reference>;

/**
 * The information neccessary to build a DocLink.
 */
export interface Reference {
  docId: string;
  section: string;
  text: string;
}
```

### ScopeApi

Scope API is essentially just grouping an array of ApiDeclarations into different categories that makes building the mdx files from a
single json file easier.

```ts
export interface ScopeApi {
  setup?: ApiDeclaration;
  start?: ApiDeclaration;
  classes: ApiDeclaration[];
  functions: ApiDeclaration[];
  interfaces: ApiDeclaration[];
  objects: ApiDeclaration[];
  enums: ApiDeclaration[];
  misc: ApiDeclaration[];
  // We may add more here as we sit fit to pull out of `misc`.
}
```

With this structure, the mdx files end up looking like:

```
### Start
<DocDefinitionList data={[actionsJson.server.start]}/>
### Functions
<DocDefinitionList data={actionsJson.server.functions}/>
### Interfaces
<DocDefinitionList data={actionsJson.server.interfaces}/>
```

### PluginApi

> A plugins API is the component that is serialized into the json file. It is broken into public, server and common components.  `serviceFolders` is a way for the system to

Reviewer 2: super super nit
```suggestion
A plugin API is the component that is serialized into the json file. It is broken into public, server and common components.  `serviceFolders` is a way for the system to
```

write separate mdx files depending on where each declaration is defined. This is because certain plugins (and core)
are huge, and can't be rendered in a single page.

```ts
export interface PluginApi {
  id: string;
  serviceFolders?: readonly string[];
  client: ScopeApi;
  server: ScopeApi;
  common: ScopeApi;
}
```

> ## Manifest file changes

Reviewer 2: Do you think it'd add more clarity if we referred to this as the "`kibana.json` Manifest file", instead of just the "Manifest file"?

Author: Yes, good idea.

**serviceFolders?: string[]**

Used by the system to group services into sub-pages. Some plugins, like data and core, have such huge APIs they are very slow to contain in a single page, and they are less consummable by solution developers. The addition of an optional list of services folders will cause the system to automatically create a separate page with every API that is defined within that folder. The caveat is that core will need to define a manifest file in order to define its service folders...

> **Using a kibana.json file for core**

Reviewer 2: nit: It feels like this section shouldn't be a sibling to `serviceFolders` and `teamOwner`, perhaps we should just put it right under the `## Manifest file changes` header?

# Future features

## Indexing stats

Can we index statistics about our API as part of this system? For example, I'm dumping information about which api declarations are missing comments in the console.

## Longer term approach to "plugin service folders"

Using sub folders is a short term plan. A long term plan hasn't been established yet, but it should fit in with our folder structure hierarchy goals, along with
any support we have for sharing services among a related set of plugins, that are not exposed as part of the public API.
# Recommendations for writing comments

## @link comments for the referenced type

Core has a pattern of writing comments like this:

```ts
  /** {@link IUiSettingsClient} */
  uiSettings: IUiSettingsClient;
```

I don't see the value in this. In the IDE, I can click on the IUiSettingsClient type and get directed there, and in the API doc system, the
type will already be clickable. This ends up with a weird looking API:

![image](../images/repeat_type_links.png)

The plan is to make @link comments work like links, which means this is unneccessary information.

I propose we avoid this kind of pattern.

## Export every referenced type

> The docs system handles broken link warnings but to avoid breaking the ci, I suggest we turn this off initially. This will mean however,

Reviewer 2: nit
```suggestion
The docs system handles broken link warnings but to avoid breaking CI, I suggest we turn this off initially. However, this will mean
```

that we may miss situations where we are referencing a type that is not actually exported. This will cause a broken link in the docs
system

For example if your index.ts file has:
```ts
export type foo: string | AnInterface;
```

and does not also export `AnInterface`, this will be a broken link in the docs system.

> Until we have better ci tools to catch these mistakes, developers will need to export every referenced type.

Reviewer 2: nit
```suggestion
Until we have better CI tools to catch these mistakes, developers will need to export every referenced type.
```

## Avoid `Pick` pattern

Connected to the above, if you use `Pick`, there are two problems. One is that it's difficult for a developer to see the functionality
available to them at a glance, since they would have to keep flipping from the interface definition to the properties that have been picked.

The second potential problem is that you will have to export the referenced type, and in some situations, it's an internal type that isn't exported.

![image](../images/api_doc_pick.png)

# Open questions

## Required attribute

`isRequired` is an optional parameter that can be used to display a badge next to the API.
We can mark function parameters that do not use `?` or `| undefined` as required. Open questions:

1. Are we okay with a badge showing for `required` rather than `optional` when marking a parameter as optional is extra work for a developer, and `required` is the default?

> 2. Should we only mark function parameters as `required` or interface/class parameters? Essentially, should any declaration that is not nullable

Reviewer 2: IMO, we should treat "required" as the default and "optional" as the exception. It feels a lot more common for parameters, etc. to be required rather than optional.

Author: ++ (cc @snide @glitteringkatie )

have the `required` tag?

## Signatures on primitive types

> 1. Should we _always_ include a signature for variables and parameters, even if they are a repeat of the TypeKind? For example:

Reviewer 2: It seems like overkill to me...

![image](../images/repeat_primitive_signature.png)

> 2. If no, should we include signatures when the only difference is `| undefined`? For function parameters this information is captured by

Reviewer 2: If we were to show a badge for "optional" as opposed to "required", it seems like it'd solve this issue, right?

Author: Yea, I proposed that to @snide but if I remember correctly, he wanted us to use `required` for consistency with API docs. I believe this is something we can iterate on, not set in stone.

the absence of the `required` badge. Is this obvious? What about class members/interface props?

## Out of scope

### REST API

This RFC does not cover REST API documentation, though it worth considering where

> REST APIs registered by plugins should go in the docs. The docs team has a proposal for this but it is not inside the `Kibana Developer Docs` mission.

Reviewer 2: Plugins shouldn't be consuming other plugin's HTTP APIs directly, right? Instead, it should be plugin to plugin communication, which might then go over HTTP...

Author: I don't think we have that written down as a recommendation anywhere.  I believe there are some cases where there is no plugin API so devs go through the HTTP API (I think some may do this with the short URL feature, or at least ML was looking into it).

Either way, I think you are saying they shouldn't go in these developer docs? They are for developers... but not necessarily developers writing Kibana plugins. So I think I'd be on board with that.

Reviewer 2: > Either way, I think you are saying they shouldn't go in these developer docs? They are for developers... but not necessarily developers writing Kibana plugins. So I think I'd be on board with that.

Yeah, they're potentially for developers, but they're also for the general public to consume as well.
